### PR TITLE
Fix heatmap rendering when zsmooth=fast

### PR DESCRIPTION
**Pull request by lvlte**
Fixes [#6564](https://github.com/plotly/plotly.js/issues/6564)

All details are in the issue description. I will add a markdown log file tomorrow.

### CODE REVIEW

**Comment by archmoj** [comment]
Thanks for the PR. 
To compare here is the updated codepen based on your PR:
[After](https://codepen.io/MojtabaSamimi/pen/jOeWamd) vs [Before](https://codepen.io/lvlte/pen/GRYpPVp).

**Comment by lvlte** [comment]
Ok, I will try to cover both the edge case (when the bug can be visualized) and the other case (invisible, which affects performance). 

Do you know by the way why the build-and-test checks failed ? It seems to fail randomly on "webgl-jasmine" or "flaky-no-gl-jasmine" from what I saw (one succeeds, the other fails) but unless I missed something it should not be related to this PR.

**Comment by archmoj** [comment]
Thanks. Those failures are common and are not related to your PR. I will rerun those two tests for you.

**Comment by lvlte** [comment]
Ok thank you. 

To better illustrate what I mean by "edge case" (visual bug) vs "other case" (performance issue) : most of the time (when `m*n` is less than `imageWidth*imageHeight`) there will be a lot of dead pixels drawn *outside* the canvas, hence this is invisible but impacts performance :

To "see" it, on current master :
```
// if(zsmooth === 'fast') {
//     canvasW = n;
//     canvasH = m;
// } else {
//     canvasW = imageWidth;
//     canvasH = imageHeight;
// }
canvasW = imageWidth;
canvasH = imageHeight;
```
Inspecting the final image of "zsmooth_methods" mock reveals how many (181x170 instead of just 5x3) pixels were drawn : 

InkedScreenshot 2023-04-13 182904

In fact I was trying to find a way to improve heatmap rendering performances both with and without zsmooth and I land on this bug :).

File: draftlogs/6565_fix.md [hunk 1]
```diff
@@ -0,0 +1 @@
+ - Fix Heatmap rendering bug and improve performance when zsmooth="fast" [[#6565](https://github.com/plotly/plotly.js/pull/6565)]
```
Comment (archmoj) [inline draftlogs/6565_fix.md:1]:
```suggestion
 - Fix heatmap rendering bug and improve performance when `zsmooth` is set to "fast" [[#6565](https://github.com/plotly/plotly.js/pull/6565)], with thanks to @lvlte for the contribution!
 ```

**Comment by archmoj** [comment]
Currently we have a bug in our image test system. 
To avoid that please rename your new mock & baseline png (as well as in the jasmine require) from 
`heatmap_small_layout_zsmooth_fast` to ``zz-heatmap_small_layout_zsmooth_fast`.
That way the order of baseline creation won't change in parallel runs and we should be good to go.
Thank you!

**Comment by archmoj** [comment]
Thanks very much for the great fix and tests! :1st_place_medal: 
:dancer:
